File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/between_two_sets.go
// https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var n1, n2 int
    _, _ = fmt.Scan(&n1, &n2)
    a := readNumbers(n1)
    b := readNumbers(n2)
    fmt.Println(betweenTwoSets(a, b))
}

func readNumbers(n int) []int {
    arr := make([]int, n)
    for i := range n {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

func betweenTwoSets(a, b []int) int {
    lcmOfA := reduce(a, lcm)
    gcdOfB := reduce(b, gcd)

    count := 0
    for i := lcmOfA; i <= gcdOfB; i += lcmOfA {
        if gcdOfB%i == 0 {
            count++
        }
    }
    return count
}

func reduce(arr []int, function func(int, int) int) int {
    res := arr[0]
    for _, x := range arr[1:] {
        res = function(res, x)
    }
    return res
}

func lcm(a, b int) int {
    return (a * b) / gcd(a, b)
}

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a%b
    }
    return a
}

File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/between_two_sets.py
# https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true

import math
from functools import reduce


def main():
    input()
    a = read_numbers()
    b = read_numbers()
    print(between_two_sets(a, b))


def read_numbers():
    return list(map(int, input().split()))


def between_two_sets(a, b):
    lcm_of_a = reduce(math.lcm, a)
    gcd_of_b = reduce(math.gcd, b)

    count = 0
    for i in range(lcm_of_a, gcd_of_b + 1, lcm_of_a):
        if gcd_of_b % i == 0:
            count += 1
    return count


if __name__ == "__main__":
    main()
